RealmSwift/SyncSubscription.swift

```diff
@@ -0,0 +1,459 @@
+        rlmSyncSubscriptionSet.removeAllSubscriptions(withClassName: type.className())
+    }
+
+    // MARK: Subscription Retrieval
+
+    /**
+     Returns the subscription at the given `position`.
+
+     - parameter position: The index for the resulting subscription.
+     */
+    public subscript(position: Int) -> SyncSubscription? {
+        throwForNegativeIndex(position)
+        return rlmSyncSubscriptionSet.object(at: UInt(position)).map { SyncSubscription($0) }
+    }
+
+    /// Returns the first object in the SyncSubscription list, or `nil` if the subscriptions are empty.
+    public var first: SyncSubscription? {
+        return rlmSyncSubscriptionSet.firstObject().map { SyncSubscription($0) }
+    }
+
+    /// Returns the last object in the SyncSubscription list, or `nil` if the subscriptions are empty.
+    public var last: SyncSubscription? {
+        return rlmSyncSubscriptionSet.lastObject().map { SyncSubscription($0) }
+    }
+}
+
+extension SyncSubscriptionSet: Sequence {
+    // MARK: Sequence Support
+
+    /// Returns a `SyncSubscriptionSetIterator` that yields successive elements in the subscription collection.
+    public func makeIterator() -> SyncSubscriptionSetIterator {
+        return SyncSubscriptionSetIterator(rlmSyncSubscriptionSet)
+    }
+}
+
+/**
+ This struct enables sequence-style enumeration for `SyncSubscriptionSet`.
+ */
+@frozen public struct SyncSubscriptionSetIterator: IteratorProtocol {
+    private let rlmSubscriptionSet: RLMSyncSubscriptionSet
+    private var index: Int = -1
+
+    init(_ rlmSubscriptionSet: RLMSyncSubscriptionSet) {
+        self.rlmSubscriptionSet = rlmSubscriptionSet
+    }
+
+    private func nextIndex(for index: Int?) -> Int? {
+        if let index = index, index < self.rlmSubscriptionSet.count - 1 {
+            return index + 1
+        }
+        return nil
+    }
+
+    mutating public func next() -> RLMSyncSubscription? {
+        if let index = self.nextIndex(for: self.index) {
+            self.index = index
+            return rlmSubscriptionSet.object(at: UInt(index))
+        }
+        return nil
+    }
+}
+
+#if swift(>=5.5) && canImport(_Concurrency)
+@available(macOS 12.0, tvOS 15.0, iOS 15.0, watchOS 8.0, *)
+extension SyncSubscriptionSet {
+    /**
+     Asynchronously creates and commit a write transaction and updates the subscription set,
+     this will not wait for the server to acknowledge and see all the data associated with this
+     collection of subscription.
+
+     - parameter block: The block containing the subscriptions transactions to perform.
+
+     - throws: An `NSError` if the transaction could not be completed successfully.
+               If `block` throws, the function throws the propagated `ErrorType` instead.
+     */
+    public func write(_ block: (() -> Void)) async throws {
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
+            lock.lock()
```

Comment:
Locking here doesn't really do much. The most immediate problem is that every entry point needs locking, not just one, or calling `await subscriptions.write()` and then continuing to use the `subscriptions` object on the calling thread will break. Adding that locking still wouldn't work, though, because RLMSyncSubscriptionSet is ultimately backed by a thread-confined Realm, and so accessing it from other threads is illegal. Each of RLMSyncSubscriptionSet's public entry points really should be calling `[_realm verifyThread]`.

This currently will often work by coincidence because `withCheckedThrowingContinuation()` doesn't _guarantee_ that it hops to a different thread, but as soon as it does it'll break.

The simplest way to provide something that'll work for now is to just make it `@MainActor` and only support using this from the main thread. A better solution might be to re-obtain the Realm and SyncSubscriptionSet from within the continuation to get a valid thread-local one and pass it to the user's block.

Reply:
I can just make this to run only on the main thread for today's release and fix this for a next release.
If we add the `@MainActor` annotation this will mean that everything we are doing on the writeOnComplete method will block the main thread, because the only thing I'm worried is if we are waiting for all the data to be synced and during this time the main thread is blocked. @tgoyne